File: app/src/main/java/com/kimjisub/launchpad/ImportPackByUrlActivity.java
package com.kimjisub.launchpad;

import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.support.v4.app.NotificationCompat;

import com.kimjisub.launchpad.api.unipad.UniPadApi;
import com.kimjisub.launchpad.api.unipad.vo.UnishareVO;
import com.kimjisub.launchpad.databinding.ActivityImportpackBinding;
import com.kimjisub.launchpad.manager.FileManager;
import com.kimjisub.launchpad.manager.Log;
import com.kimjisub.launchpad.manager.NotificationManager;
import com.kimjisub.launchpad.manager.Unipack;
import com.kimjisub.launchpad.network.UnishareDownloader;

import java.io.File;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ImportPackByUrlActivity extends BaseActivity {
	ActivityImportpackBinding b;

	String code;

	NotificationCompat.Builder notificationBuilder;
	int id;
	long fileSize;
	long downloadedSize;
	String identifyCode = "";
	String errorMsg = "";
	UnishareVO unishare;
	Unipack unipack;
	Throwable throwable;
	int prevPercent = 0;

	void initVar() {
		code = getIntent().getData().getQueryParameter("code");
		log("code: " + code);

		id = (int) (Math.random() * Integer.MAX_VALUE);

		notificationBuilder = new NotificationCompat.Builder(ImportPackByUrlActivity.this)
				.setAutoCancel(true)
				.setSmallIcon(R.mipmap.ic_launcher);
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
			notificationBuilder.setChannelId(NotificationManager.Channel.DOWNLOAD);
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		b = setContentViewBind(R.layout.activity_importpack);
		initVar();

		identifyCode = "#" + code;
		setStatus(Status.prepare);

		UniPadApi.getService().unishare_get(code).enqueue(new Callback<UnishareVO>() {
			@Override
			public void onResponse(Call<UnishareVO> call, Response<UnishareVO> response) {
				if (response.isSuccessful()) {
					unishare = response.body();
					log("title: " + unishare.title);
					log("producerName: " + unishare.producer);
					identifyCode = unishare.title + " #" + unishare._id;
					setStatus(Status.getInfo);

					new UnishareDownloader(unishare, F_UniPackRootExt, new UnishareDownloader.UnishareDownloadListener() {
						@Override
						public void onDownloadStart() {
							log("Download start");
							setStatus(Status.downloadStart);
						}

						@Override
						public void onGetFileSize(long fileSize_) {
							fileSize = fileSize_ > 0 ? fileSize_ : unishare.fileSize;
							log("fileSize: " + fileSize_ + " → " + fileSize);
						}

						@Override
						public void onDownloading(long downloadedSize_) {
							downloadedSize = downloadedSize_;

							setStatus(Status.downloading);
						}

						@Override
						public void onDownloadEnd(File zip) {
							log("Download End");
						}

						@Override
						public void onAnalyzeStart() {
							log("Analyzing Start");
							setStatus(Status.analyzing);
						}

						@Override
						public void onAnalyzeSuccess(Unipack unipack_) {
							log("Analyzing Success");
							unipack = unipack_;
							setStatus(Status.success);
						}

						@Override
						public void onAnalyzeFail(Unipack unipack) {
							log("Analyzing Fail");
							Log.err(unipack.ErrorDetail);
							errorMsg = unipack.ErrorDetail;
							setStatus(Status.failed);
						}

						@Override
						public void onAnalyzeEnd(File folder) {
							log("Analyzing End");
							delayFinish();
						}

						@Override
						public void onException(Throwable throwable_) {
							throwable = throwable_;
							throwable.printStackTrace();
							setStatus(Status.exception);
							log("Exception: " + throwable.getMessage());
						}
					});
				} else {
					switch (response.code()) {
						case 404:
							log("404 Not Found");
							setStatus(Status.notFound);//, "Not Found");
							break;
					}
				}
			}

			@Override
			public void onFailure(Call<UnishareVO> call, Throwable t) {
				log("server error");
				setStatus(Status.failed);//, "server error\n" + t.getMessage());
			}
		});
	}


	enum Status {
		prepare(0, R.string.wait_a_sec),
		getInfo(1, R.string.wait_a_sec),
		downloadStart(2, R.string.downloadWaiting),
		downloading(3, R.string.downloading),
		analyzing(4, R.string.analyzing),
		success(5, R.string.success, false),
		failed(6, R.string.failed, false),
		exception(7, R.string.exceptionOccurred, false),
		notFound(8, R.string.unipackNotFound, false);

		int value;
		int titleStringId;
		boolean ongoing = true;

		Status(int value, int titleStringId) {
			this.value = value;
			this.titleStringId = titleStringId;
		}

		Status(int value, int titleStringId, boolean ongoing) {
			this.value = value;
			this.titleStringId = titleStringId;
			this.ongoing = ongoing;
		}
	}


	void setStatus(Status status) {
		runOnUiThread(() -> {
			switch (status) {
				case prepare:
					b.title.setText(status.titleStringId);
					b.message.setText(identifyCode);
					notificationBuilder.setContentTitle(identifyCode);
					notificationBuilder.setContentText(lang(status.titleStringId));
					break;
				case getInfo:
					b.title.setText(status.titleStringId);
					b.message.setText(identifyCode);
					notificationBuilder.setContentTitle(identifyCode);
					notificationBuilder.setContentText(lang(status.titleStringId));
					break;
				case downloadStart:
					b.title.setText(status.titleStringId);
					b.message.setText("#" + code + "\n" + unishare.title + "\n" + unishare.producer);
					notificationBuilder.setContentTitle(identifyCode);
					notificationBuilder.setContentText(lang(status.titleStringId));
					notificationBuilder.setProgress(100, 0, true);
					break;
				case downloading:
					int percent = (int) ((float) downloadedSize / fileSize * 100);
					String downloadedSizeMB = FileManager.byteToMB(downloadedSize);
					String fileSizeMB = FileManager.byteToMB(fileSize);

					if (prevPercent == percent) return;
					prevPercent = percent;

					b.title.setText(status.titleStringId);
					b.message.setText(percent + "%\n" + downloadedSizeMB + " / " + fileSizeMB + "MB");
					notificationBuilder.setContentTitle(identifyCode);
					notificationBuilder.setContentText(percent + "%\n" + downloadedSizeMB + " / " + fileSizeMB + "MB");
					notificationBuilder.setProgress(100, percent, false);
					break;
				case analyzing:
					b.title.setText(status.titleStringId);
					b.message.setText("#" + code + "\n" + unishare.title + "\n" + unishare.producer);
					notificationBuilder.setContentTitle(identifyCode);
					notificationBuilder.setContentText(lang(status.titleStringId));
					notificationBuilder.setProgress(100, 0, true);
					break;
				case success:
					b.title.setText(status.titleStringId);
					b.message.setText(unipack.getInfoText(ImportPackByUrlActivity.this));
					notificationBuilder.setContentTitle(identifyCode);
					notificationBuilder.setContentText(lang(status.titleStringId));
					notificationBuilder.setProgress(0, 0, false);
					break;
				case failed:
					b.title.setText(status.titleStringId);
					b.message.setText(errorMsg);
					notificationBuilder.setContentTitle(identifyCode);
					notificationBuilder.setContentText(lang(status.titleStringId));
					notificationBuilder.setProgress(0, 0, false);
					break;
				case exception:
					b.title.setText(status.titleStringId);
					b.message.setText(throwable.getMessage());
					notificationBuilder.setContentTitle(identifyCode);
					notificationBuilder.setContentText(lang(status.titleStringId));
					notificationBuilder.setProgress(0, 0, false);
					break;
				case notFound:
					b.title.setText(status.titleStringId);
					b.message.setText("#" + code + "\n" + unishare.title + "\n" + unishare.producer);
					notificationBuilder.setContentTitle(identifyCode);
					notificationBuilder.setContentText(lang(status.titleStringId));
					notificationBuilder.setProgress(0, 0, false);
					break;
			}


			notificationBuilder.setOngoing(status.ongoing);

			NotificationManager.getManager(ImportPackByUrlActivity.this).notify(id, notificationBuilder.build());
		});
	}

	void log(String msg) {
		runOnUiThread(() -> b.info.append(msg + "\n"));
	}

	void delayFinish() {
		log("delayFinish()");
		runOnUiThread(() -> new Handler().postDelayed(() -> {
			finish();
			overridePendingTransition(R.anim.activity_in, R.anim.activity_out);
		}, 3000));
		//restartApp(this);
	}

	@Override
	public void onResume() {
		super.onResume();
	}

	@Override
	public void onDestroy() {
		super.onDestroy();
	}


}

File: app/src/main/java/com/kimjisub/launchpad/activity/ImportPackByUrlActivity.kt
package com.kimjisub.launchpad.activity

import android.os.Build
import android.os.Bundle
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.lifecycleScope
import com.kimjisub.launchpad.R.anim
import com.kimjisub.launchpad.ui.theme.Background1
import com.kimjisub.launchpad.ui.theme.Gray1
import com.kimjisub.launchpad.ui.theme.OverlayLight
import com.kimjisub.launchpad.R.string
import com.kimjisub.launchpad.api.unipad.UniPadApi.service
import com.kimjisub.launchpad.api.unipad.vo.UnishareVO
import com.kimjisub.launchpad.manager.FileManager
import com.kimjisub.launchpad.tool.Log
import com.kimjisub.launchpad.tool.UniPackDownloader
import com.kimjisub.launchpad.unipack.UniPack
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.IOException

class ImportPackByUrlActivity : BaseActivity() {
	companion object {
		private const val INSTALL_COMPLETE_DELAY_MS = 3000L

		/** Shown with the question so the person can see where the file comes from. */
		private const val UNISHARE_HOST = "api.unipad.io"
	}

	private val titleText = mutableStateOf("")
	private val messageText = mutableStateOf("")
	private val infoText = mutableStateOf("")

	/**
	 * The pack a deep link wants to install, held while the user decides.
	 *
	 * A `unipad://` link used to go straight from the metadata response into the
	 * download. Anything that can open a link on this device could therefore make
	 * the app fetch and unpack a file with no say from the person holding it. The
	 * metadata call already ran by this point, so the question can name the pack
	 * and its producer instead of asking about an anonymous download.
	 */
	private val pendingInstall = mutableStateOf<UnishareVO?>(null)

	private val code: String? by lazy { intent?.data?.getQueryParameter("code") }

	private var identifyCode = ""

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		titleText.value = getString(string.wait_a_sec)

		log("code: $code")
		identifyCode = getString(string.import_pack_code_format, code.orEmpty())
		messageText.value = identifyCode

		setContent {
			val title by titleText
			val message by messageText
			val info by infoText
			val pending by pendingInstall
			ImportPackScreen(
				titleText = title,
				messageText = message,
				infoText = info,
				confirming = pending != null,
				onAccept = { pending?.let { startInstall(it) } },
				onCancel = { finish() },
			)
		}

		val requestCode = code ?: return
		lifecycleScope.launch {
			try {
				val response = withContext(Dispatchers.IO) {
					service.getUnishare(requestCode)
				}
				if (response.isSuccessful) {
					val unishare = response.body() ?: return@launch
					identifyCode = getString(string.import_pack_title_format, unishare.title, unishare._id)
					log("title: ${unishare.title}")
					log("producerName: ${unishare.producer}")

					// Ask first. The download starts from onAccept, not from here.
					titleText.value = unishare.title.orEmpty()
					messageText.value = unishare.producer.orEmpty()
					infoText.value = getString(string.import_pack_confirm_source, UNISHARE_HOST)
					pendingInstall.value = unishare
				} else {
					when (response.code()) {
						404 -> {
							log("404 Not Found")
							titleText.value = getString(string.unipackNotFound)
							messageText.value = identifyCode
						}
					}
				}
			} catch (e: IOException) {
				log("server error")
				Log.err("Unishare API call failed", e)
				titleText.value = getString(string.failed)
				messageText.value = getString(string.server_error_format, e.message.orEmpty())
			}
		}
	}

	fun startInstall(unishare: UnishareVO) {
		pendingInstall.value = null
		val packTitle = getString(string.import_pack_title_format, unishare.title, unishare._id)
		UniPackDownloader(
			context = this,
			title = packTitle,
			url = "https://api.unipad.io/unishare/${unishare._id}/download",
			workspace = ws.downloadWorkspace.file,
			folderName = packTitle,
			listener = object : UniPackDownloader.Listener {
				override fun onInstallStart() {
					log("Install start")
					titleText.value = getString(string.downloadWaiting)
					messageText.value = getString(string.import_pack_info_format, code.orEmpty(), unishare.title, unishare.producer)
				}

				override fun onGetFileSize(
					fileSize: Long,
					contentLength: Long,
					preKnownFileSize: Long,
				) {
					log("fileSize: $contentLength → $fileSize")
				}

				override fun onDownloadProgress(
					percent: Int,
					downloadedSize: Long,
					fileSize: Long,
				) {
				}

				override fun onDownloadProgressPercent(
					percent: Int,
					downloadedSize: Long,
					fileSize: Long,
				) {
					val downloadedMB: String = FileManager.byteToMB(downloadedSize)
					val fileSizeMB: String = FileManager.byteToMB(fileSize)

					titleText.value = getString(string.downloading)
					messageText.value = getString(string.download_progress_format, percent, downloadedMB, fileSizeMB)
				}

				override fun onImportStart(zip: File) {
					log("Import Start")
					titleText.value = getString(string.importing)
					messageText.value = getString(string.import_pack_info_format, code.orEmpty(), unishare.title, unishare.producer)
				}

				override fun onInstallComplete(folder: File, unipack: UniPack) {
					log("Install Success")
					titleText.value = getString(string.success)
					messageText.value = unipack.infoToString(this@ImportPackByUrlActivity)
					delayFinish()
				}

				override fun onException(throwable: Throwable) {
					log("Exception: ${throwable.message}")
					Log.err("ImportPackByUrl download failed", throwable)
					titleText.value = getString(string.exceptionOccurred)
					messageText.value = throwable.message ?: ""
				}
			},
			scope = lifecycleScope,
		)
	}

	private fun log(msg: String) {
		infoText.value += msg + "\n"
		Log.download(msg)
	}

	private fun delayFinish() {
		log("delayFinish()")
		lifecycleScope.launch {
			delay(INSTALL_COMPLETE_DELAY_MS)

			finish()
			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
				overrideActivityTransition(OVERRIDE_TRANSITION_CLOSE, anim.activity_in, anim.activity_out)
			} else {
				@Suppress("DEPRECATION")
				overridePendingTransition(anim.activity_in, anim.activity_out)
			}
		}
	}
}

private val grayColor = Gray1

@Composable
private fun ImportPackScreen(
	titleText: String,
	messageText: String,
	infoText: String,
	confirming: Boolean = false,
	onAccept: () -> Unit = {},
	onCancel: () -> Unit = {},
) {
	Box(
		modifier = Modifier
			.fillMaxSize()
			.background(OverlayLight),
		contentAlignment = Alignment.Center,
	) {
		Box(
			modifier = Modifier
				.width(300.dp)
				.height(200.dp)
				.background(Background1),
		) {
			if (infoText.isNotEmpty()) {
				Text(
					text = infoText,
					color = grayColor,
					fontSize = 12.sp,
					modifier = Modifier
						.align(Alignment.TopStart)
						.alpha(0.5f),
				)
			}

			Column(
				modifier = Modifier
					.align(Alignment.Center)
					.verticalScroll(rememberScrollState()),
				horizontalAlignment = Alignment.CenterHorizontally,
			) {
				Text(
					text = titleText,
					color = grayColor,
					fontSize = 25.sp,
				)
				Text(
					text = messageText,
					color = grayColor,
					fontSize = 18.sp,
					textAlign = TextAlign.Center,
				)

				if (confirming) {
					Text(
						text = stringResource(string.import_pack_confirm),
						color = grayColor,
						fontSize = 14.sp,
						textAlign = TextAlign.Center,
						modifier = Modifier.padding(top = 12.dp),
					)
					Row(
						modifier = Modifier.padding(top = 8.dp),
						horizontalArrangement = Arrangement.spacedBy(8.dp),
					) {
						TextButton(onClick = onCancel) {
							Text(text = stringResource(string.cancel), color = grayColor)
						}
						TextButton(onClick = onAccept) {
							Text(text = stringResource(string.accept), color = grayColor)
						}
					}
				}
			}
		}
	}
}
